File: ruby/pokemon/pokemon.rb
# frozen_string_literal: true

# Rubyは抽象クラスなし
class Pokemon
  attr_accessor :name, :type1, :type2, :hp, :mp

  def initialize(name, type1, type2, hp, mp)
    @name = name
    @type1 = type1
    @type2 = type2
    @hp = hp
    @mp = mp
  end

  def attack
    "#{@name}のこうげき!"
  end
end

# Rubyは単一継承
class Pikachu < Pokemon
  attr_accessor :type3

  def initialize(name, type1, type2, hp, mp, type3)
    super(name, type1, type2, hp, mp)
    @type3 = type3
  end

  def attack
    "#{super}\n#{@name}の10万ボルト!"
  end

  private

  def deffence
    "#{super}\n#{@name}の身を守る!"
  end
end

# 多重継承の仕組みを作る
module Loggable
  def log(text)
    p "[LOG] #{text}"
  end
end

# 多重継承の仕組みを作る
module Pokemon2
  attr_accessor :name, :type1, :type2, :hp, :mp

  def initialize(name, type1, type2, hp, mp)
    @name = name
    @type1 = type1
    @type2 = type2
    @hp = hp
    @mp = mp
  end

  def attack
    "#{@name}のこうげき!"
  end
end

# 多重継承の仕組みを作る
class Zenigame
  include Loggable
  include Pokemon2

  attr_accessor :type3

  def initialize(name, type1, type2, hp, mp, type3)
    super(name, type1, type2, hp, mp)
    @type3 = type3
  end

  def attack
    "#{super}\n#{@name}のみずでっぽう"
  end

  def title
    log 'title is called.'
    'Pokemon'
  end
end

pokemons = (1..100).to_a.map { Pokemon.new(:リザードン, :ほのお, :ひこう, 100, 10) }
# p pokemons.size
p pokemons[0].name
# p pokemons[0].type1
# p pokemons[0].type2
p pokemons[0].attack
p '------------------'

pika = Pikachu.new(:ピカチュウ, :でんき, :とぶ, 50, 5, :むし)
p pika.name
puts pika.attack
p pika.type3
# p pika.deffence # 呼び出し不可
p '------------------'

zeni = Zenigame.new(:ぜにがめ, :みず, :およぐ, 80, 8, :かめ)
p zeni.title
puts zeni.attack
p '------------------'

File: python/pokemon.py
# Pythonにもデストラクタはある（ほぼ使ったことがない）
class PokemonDel:
    def __init__(self, name):
        self.name = name

    def __del__(self):
        print(f"{self.name}オブジェクトが破棄されました")

# オブジェクトの作成
pokemon = PokemonDel("リザードン")

# オブジェクトの削除
del pokemon

print("-------------------------------------")
# 継承とポリモーフィズム(同じメソッドで異なる振る舞いをすること)
class Pokemon:
    def __init__(self, name, type1, type2, hp):
        self.name = name
        self.type1 = type1
        self.type2 = type2
        self.hp = hp
    
    def attack(self):
        print(f"{self.name}のこうげき")

class Pikachu(Pokemon):
    def __init__(self, name, type1, type2, hp):
        super().__init__(name, type1, type2, hp)

    def attack(self):
        # super().attack()
        print(f"{self.name}のかいしんのいちげき")

poke = Pikachu("ピカチュウ", "でんき", "むし", "100")
print(poke.name)
poke.attack()

print("-------------------------------------")
# 抽象クラス
# 抽象クラスを継承したクラスは全ての抽象メソッドをオーバーライドしなければならない
from abc import ABC, abstractmethod

class Pokemon(ABC): # ABCクラスの継承

    @abstractmethod # デコレーターを付けて抽象メソッドを定義
    def attack(self):
        pass # passとする

class Zenigame(Pokemon):
    def attack(self): # ここでオーバーライド
        print("みすでっぽう")

poke = Zenigame() # オーバーライドしないとエラーになる
poke.attack()

# ※Pythonにインタフェースを定義するための固有の機能はないが、同じことをクラスを使って実現は可能

print("-------------------------------------")
# カプセル化
class Pokemon:
    def __init__(self):
        self.__name = "myu"

    def __attack(self):
        print("はかいこうせん")

poke = Pokemon()
# poke.__name # アクセス不可
# poke.__attack() # アクセス不可